kernel/thread/worker.c

```diff
@@ -0,0 +1,89 @@
+/* KallistiOS ##version##
+
+   worker.c
+   Copyright (C) 2024 Paul Cercueil
+*/
+
+#include <kos/genwait.h>
+#include <kos/thread.h>
+#include <kos/worker_thread.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+struct kthread_worker {
+    kthread_t *thd;
+    void (*routine)(void *);
+    void *data;
+    bool pending;
+    bool quit;
+};
+
+static void * thd_worker_thread(void *d) {
+    kthread_worker_t *worker = d;
+    uint32_t flags;
+
+    for (;;) {
+        flags = irq_disable();
+
+        if (!worker->pending)
+            genwait_wait(worker, worker->thd->label, 0, NULL);
+
+        irq_restore(flags);
+
+        if (worker->quit)
+            break;
+
+        worker->pending = false;
+        worker->routine(worker->data);
+    }
+
+    return NULL;
+}
+
+kthread_worker_t * thd_worker_create(void (*routine)(void *), void *param) {
```

Comment:
I'm not sure how this should get factored out, but I think it would make a lot of sense to have a version of this as well (perhaps `thd_worker_create_ex`) that would take in a `kthread_attr_t` to define detailed params. Since the worker is only run on demand I could see though it makes sense if the intent was that all those could be manipulated after the worker is created.

Reply:
I can do that, if you think that's needed. Currently the attributes can be added after the worker is created, yes.

Reply:
Ah, that would be a good idea, actually. For consistency and orthogonality and shit.

Reply:
@gyrovorbis it's done in the new version.